### DISCUSSION TITLE
Investigation: SubmarineInfo changes to shorten load times and lower memory usage.

### DISCUSSION THREAD

**Discussion by someone972** [FakeFishGames/Barotrauma#4893] [Code questions and modding support]
Not sure if 'Code questions' is the right topic for this, but since this was a code investigation it seemed appropriate. This is just the results of an investigation I was doing. This whole "feature" might not even be necessary with the new feature of deleting downloaded subs each launch. Feel free to leave questions or feedback on how useful this would be, I don't think I'll be pursuing it further unless there's interest.

### The issue: 
Barotrauma client and server loads **all** submarines into memory at the start of the program. This includes a preview image which is only used in a couple places, and the full submarine XML. This is obviously a waste of memory, and with large amounts of subs this can significantly slow down loading and cause issues launching servers.

### Investigation:
Investigated reworking the sub file to mitigate these issues. Idea was to split the single compressed XML into a ZipArchive of 3 (or more?) files. 
- preview.png - The preview image for the sub, in regular PNG format. Since it's not in base64 encoding this saves a bit of space.
- metadata.xml - Move a small subset of information out of the submarine XML. This would be the information needed for display in the GUI for selection, such as name, description, price, etc.
- submarine.xml - The rest of the submarine data. This is all the data which is needed to actually spawn and draw the sub.

The benefit of this would be that each file can be decompressed individually. This allows for lazy loading of the preview and main sub data. The initial loading routine would load only the sub metadata out of the archive, which is a very small subset of data and should be both quick to load and low on memory usage. For the preview image, in the couple of places where it's used, it could be asynchronously loaded out of the archive at the time it's needed (potentially with caching), and all the previews could be disposed of once the sub selection is no longer needed. The main sub XML would only need to be loaded for submarines which are actually being spawned, for an additional memory saving.

### Findings:
- In GameSession, it's often confusing what is using the SubmarineInfo, the Submarine, or Submarine.MainSub. No single source of truth.
- It would be fairly trivial to convert all the subs to the new format on first load, but I wonder how repeatable ZipArchive is. Could potentially run into issues with the MD5 hash if two people converted the same sub but the archive was slightly different, despite decompressing to identical data. Would require investigation, and potentially a different hashing method.
- Netcode for sending subs would probably be fine as-is.
- Determining which format a sub file is in would be trivial by comparing magic number, even if the same .sub extension is used.
- SubmarineInfo is initialized in a lot of places from several constructors, seems like there's something that could be cleaned up here.

### Summary:
Although there would be some tricky parts, this would be possible to implement. I ran into some difficulty stemming from not knowing what state is available when, but that's a different issue and is more of a code knowledge issue than a blocking issue. However, this whole change might not be necessary given the removal of downloaded subs. Unless there's actually a need/desire for this functionality, will likely drop this.

**Comment by Leonidas-ai** [comment]
I'm glad someone is tackling this.
